functions/index.js

```diff
@@ -76,23 +83,54 @@ exports.fetch = functions.https.onRequest(async (req, res) => {
 	res.json({secret: secret.data().secret});
 });
 
+function decrementAccessesLeft(id) {
+	admin.firestore().collection('messages').doc(id).update({accessesLeft: DECREMENT})
+}
+
 /**
-* Deletes a secret by id.
 *
-* params:
-*	- id
+* Deletes all messages that have run out of accesses or time expired
+*
 */
-exports.delete = functions.https.onRequest(async (req, res) => {
-	const id = req.query.id;
+exports.deleteExpiredMessages = functions.https.onRequest(async (req, res) => {
+  if (req.method !== "DELETE") {
+    return res.sendStatus(405);
+  }
 
-	if (typeof id !== "string") {
-		return res.sendStatus(400);
-	}
-	
-	const result = admin.firestore().collection('messages').doc(id).delete();
-	res.json({result: "ok"});
+  await _deleteExpiredMessages();
+
+  res.json({result: "ok"});
 });
 
-function decrementAccessesLeft(id) {
-	admin.firestore().collection('messages').doc(id).update({accessesLeft: DECREMENT})
+async function _deleteExpiredMessages() {
+  const msgs = admin.firestore().collection('messages');
+  const noAccessesLeft = msgs.where('accessesLeft', '<=', 0).limit(BATCH_SIZE_LIMIT);
+  const timeExpired = msgs.where('expiryTime', '<', Date.now() / 1000).limit(BATCH_SIZE_LIMIT);
+
+  return Promise.all([
+  	deleteQueryBatch(noAccessesLeft), 
+  	deleteQueryBatch(timeExpired)
+  	]);
+};
+
+
+async function deleteQueryBatch(query) {
+  const snapshot = await query.get();
+
+  const batchSize = snapshot.size;
+  if (batchSize === 0) {
+    return;
+  }
+
+  // Delete current batch of documents
+  const batch = admin.firestore().batch();
+  snapshot.docs.forEach((doc) => {
+    batch.delete(doc.ref);
+  });
+  await batch.commit();
+
+  // Repeat
+  process.nextTick(() => {
+    deleteQueryBatch(query);
+  });
```

Comment:
Curious as to what line 133 and line 134 is doing. From my understanding of the deleteQueryBatch method, we pass a reference to a collection of messages where the # of access is <= 0 to the _deleteQueryBatch_ method.
Then we proceed to delete all the message objects from the batch.

batch.commit() is then called to commit all pending operations to the database.

Not sure why we need to call deleteQueryBatch recursively on line 134.

Thanks!

Reply:
The base case of the recursive function is `batchSize === 0`

There can be multiple batches of messages to delete. The batch size has a max limit which I have defined as a new constant in this file.